clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,287 @@
+            } catch (KafkaException e) {
+                logger.error("Received unexpected error in heartbeat response: {}", e.getMessage());
+            }
+            return;
+        }
+
+        onErrorResponse(response, currentTimeMs);
+    }
+
+    private void onErrorResponse(final ConsumerGroupHeartbeatResponse response,
+                                 final long currentTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(currentTimeMs);
+        short errorCode = response.data().errorCode();
+        if (errorCode == Errors.NOT_COORDINATOR.code() || errorCode == Errors.COORDINATOR_NOT_AVAILABLE.code()) {
+            logInfo("Coordinator is either not started or not valid. Retrying", response, currentTimeMs);
+            coordinatorRequestManager.markCoordinatorUnknown(response.data().errorMessage(), currentTimeMs);
+        } else if (errorCode == Errors.COORDINATOR_LOAD_IN_PROGRESS.code()) {
+            // retry
+            logInfo("Coordinator {} is loading. Retrying", response, currentTimeMs);
+        } else {
+            onFatalErrorResponse(response);
+        }
+    }
+
+    private void onFatalErrorResponse(final ConsumerGroupHeartbeatResponse response) {
+        final short errorCode = response.data().errorCode();
+        if (errorCode == Errors.GROUP_AUTHORIZATION_FAILED.code()) {
+            GroupAuthorizationException error = GroupAuthorizationException.forGroupId(membershipManager.groupId());
+            logger.error("GroupHeartbeatRequest failed due to group authorization failure: {}", error.getMessage());
+            nonRetriableErrorHandler.handle(error);
+        } else if (errorCode == Errors.INVALID_REQUEST.code()) {
+            logger.error("GroupHeartbeatRequest failed due to fatal error: {}", response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.INVALID_REQUEST.exception());
+        } else if (errorCode == Errors.GROUP_MAX_SIZE_REACHED.code()) {
+            logger.error("GroupHeartbeatRequest failed due to the max group size limit: {}",
+                response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.GROUP_MAX_SIZE_REACHED.exception());
+        } else if (errorCode == Errors.UNSUPPORTED_ASSIGNOR.code()) {
+            logger.error("GroupHeartbeatRequest failed due to unsupported assignor {}: {}",
+                membershipManager.assignorSelection(), response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.UNSUPPORTED_ASSIGNOR.exception());
+        } else if (errorCode == Errors.UNRELEASED_INSTANCE_ID.code()) {
+            logger.error("GroupHeartbeatRequest failed due to the instance id {} was not released: {}",
+                membershipManager.groupInstanceId().orElse("null"),
+                response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.UNRELEASED_INSTANCE_ID.exception());
+            membershipManager.failMember();
```

Comment:
Can we rename `failMember` to be more descriptive?

Reply:
maybeTransitionToFailure ?